workos/event_objects/directory.py

```diff
@@ -0,0 +1,216 @@
+from typing import List, Literal, TypedDict
+from enum import Enum
+
+
+class DirectoryType(Enum):
+    AZURE_SCIM_v2 = "azure scim v2.0"
+    BAMBOO_HR = "bamboohr"
+    BREATHE_HR = "breathe hr"
+    CEZANNE_HR = "cezanne hr"
+    CYBERARK_SCIM_v2 = "cyperark scim v2.0"
+    FOURTH_HR = "fourth hr"
+    GENERIC_SCIM_v2 = "generic scim v2.0"
+    GOOGLE = "gsuite directory"
+    HIBOB = "hibob"
+    JUMPCLOUD_SCIM_v2 = "jump cloud scim v2.0"
+    OKTA_SCIM_v2 = "okta scim v2.0"
+    ONELOGIN_SCIM_v2 = "onelogin scim v2.0"
+    PEOPLE_HR = "people hr"
+    PERSONIO = "personio"
+    PINGFEDERATE_SCIM_v2 = "pingfederate scim v2.0"
+    RIPPLING_SCIM_v2 = "rippling v2.0"
+    SFTP = "sftp"
+    SFTP_WORKDAY = "sftp workday"
+    WORKDAY = "workday"
+
+
+class DirectoryState(Enum):
+    ACTIVE = "active"
+    DELETING = "deleting"
+
+
+class OrganizationDomain(TypedDict):
+    id: str
+    domain: str
+
+
+class WorkOSOrganizationDomain:
+    id: str
+    domain: str
+
+    @classmethod
+    def construct_from_response(cls, response: dict):
+        instance = cls()
+        for k, v in response.items():
+            setattr(instance, k, v)
+
+        return instance
+
+    def to_dict(self) -> OrganizationDomain:
+        return self.__dict__
+
+    def __str__(self):
+        return f"{self.__class__.__name__} {self.to_dict()}"
+
+
+class DirectoryEventWithLegacyFields(TypedDict):
+    id: str
+    name: str
+    type: DirectoryType
+    state: DirectoryState
+    domains: List[OrganizationDomain]
+    organization_id: str
+    created_at: str
+    updated_at: str
+    external_key: str
+    object: Literal["directory"]
+
+
+class WorkOSDirectoryEventWithLegacyFields:
+    id: str
+    name: str
+    type: DirectoryType
+    state: DirectoryState
+    domains: List[WorkOSOrganizationDomain]
+    organization_id: str
+    created_at: str
+    updated_at: str
+    external_key: str
+    object: Literal["directory"]
+
+    @classmethod
+    def construct_from_response(cls, response: dict):
+        instance = cls()
+        for k, v in response.items():
+            if k == "domains":
```

Comment:
not sure if there's a better way than to explicitly add a check against the key so it doesn't all just become nested dicts

Reply:
In terms of a quick implementation, this will work.

However, this also feels like we are reinventing the wheel. It's object deserialization based on a schema. This is a solved problem and there are a number of libraries that would do this for us like [Marshmallow](https://marshmallow.readthedocs.io/en/stable/) ... and [pydantic](https://github.com/samuelcolvin/pydantic), [typeguard](https://github.com/agronholm/typeguard) [typical](https://github.com/seandstewart/typical/),  or [pytypes](https://github.com/Stewori/pytypes)

Reply:
Yeah... I can see the limitations now. Nested objects are complicated and for having TypedDicts we need to repeat the attributes. I'm considering `pydantic` which is widely used. Here's what I got:

```python
from pydantic import BaseModel
from typing import TypedDict, List, Optional, Literal
from enum import Enum

class DirectoryType(Enum):
    AZURE_SCIM_v2 = "azure scim v2.0"
    BAMBOO_HR = "bamboohr"
    BREATHE_HR = "breathe hr"
    CEZANNE_HR = "cezanne hr"
    CYBERARK_SCIM_v2 = "cyperark scim v2.0"
    FOURTH_HR = "fourth hr"
    GENERIC_SCIM_v2 = "generic scim v2.0"
    GOOGLE = "gsuite directory"
    HIBOB = "hibob"
    JUMPCLOUD_SCIM_v2 = "jump cloud scim v2.0"
    OKTA_SCIM_v2 = "okta scim v2.0"
    ONELOGIN_SCIM_v2 = "onelogin scim v2.0"
    PEOPLE_HR = "people hr"
    PERSONIO = "personio"
    PINGFEDERATE_SCIM_v2 = "pingfederate scim v2.0"
    RIPPLING_SCIM_v2 = "rippling v2.0"
    SFTP = "sftp"
    SFTP_WORKDAY = "sftp workday"
    WORKDAY = "workday"


class DirectoryState(Enum):
    ACTIVE = "active"
    DELETING = "deleting"


class DirectoryEventData(BaseModel):
    id: str
    name: str
    type: DirectoryType
    state: DirectoryState
    organization_id: str
    created_at: str
    updated_at: str
    object: Literal["directory"]


response = {
    "id": "Test",
    "name": "Test",
    "type": "gsuite directory",
    "state": "active",
    "organization_id": "org_123",
    "created_at": "2020-12-30T10:00:00Z",
    "updated_at": "2020-12-30T10:00:00Z",
    "object": "directory",
}

obj = DirectoryEventData.parse_obj(response)
print(obj)

directory = DirectoryEventData.parse_obj(obj)
print(directory)

another = DirectoryEventData(**directory.model_dump())
print(another)

```

Reply:
hmm in that the "state" is still just a string directory from the json reqponse, as opposed to the enum type (which would be DirectoryState.Active) no?

Reply:
@jonatascastro12 can you paste what the output of these lines is?
```python
obj = DirectoryEventData.parse_obj(response)
print(obj)

directory = DirectoryEventData.parse_obj(obj)
print(directory)

another = DirectoryEventData(**directory.model_dump())
print(another)
```

Reply:
The output. Those enums come typed as expected 👍 . 
```
id='Test' name='Test' type=<DirectoryType.GOOGLE: 'gsuite directory'> state=<DirectoryState.ACTIVE: 'active'> domains=[OrganizationDomain(id='Test', domain='Test', object='organization_domain')] organization_id='org_123' created_at='2020-12-30T10:00:00Z' updated_at='2020-12-30T10:00:00Z' object='directory'
id='Test' name='Test' type=<DirectoryType.GOOGLE: 'gsuite directory'> state=<DirectoryState.ACTIVE: 'active'> domains=[OrganizationDomain(id='Test', domain='Test', object='organization_domain')] organization_id='org_123' created_at='2020-12-30T10:00:00Z' updated_at='2020-12-30T10:00:00Z' object='directory'
id='Test' name='Test' type=<DirectoryType.GOOGLE: 'gsuite directory'> state=<DirectoryState.ACTIVE: 'active'> domains=[OrganizationDomain(id='Test', domain='Test', object='organization_domain')] organization_id='org_123' created_at='2020-12-30T10:00:00Z' updated_at='2020-12-30T10:00:00Z' object='directory'
```

Reply:
Those are 3 equal objects. I just wanted to show ways that you can initialize, either with the same object, a dictionary or key arguments.